lib/elixir/lib/access.ex

```diff
@@ -670,49 +676,57 @@ defmodule Access do
   An error is raised if the accessed structure is not a list:
 
       iex> get_in(%{}, [Access.at(1)])
-      ** (RuntimeError) Access.at/1 expected a list, got: %{}
+      ** (RuntimeError) Access.at/2 expected a list, got: %{}
 
+  An error is raised if your default value does not implement the `Access`
+  behaviour but you try to access values on it:
+
+      iex> list = [%{name: "john"}]
+      iex> get_in(list, [Access.at(1), :name])
+      nil
+      iex> get_in(list, [Access.at(1, :my_default), :name])
+      ** (FunctionClauseError) no function clause matching in Access.get/3
   """
   @spec at(integer) :: access_fun(data :: list, get_value :: term)
-  def at(index) when is_integer(index) do
-    fn op, data, next -> at(op, data, index, next) end
+  def at(index, default \\ nil) when is_integer(index) do
+    fn op, data, next -> at(op, data, index, default, next) end
   end
 
-  defp at(:get, data, index, next) when is_list(data) do
-    data |> Enum.at(index) |> next.()
+  defp at(:get, data, index, default, next) when is_list(data) do
+    data |> Enum.at(index, default) |> next.()
   end
 
-  defp at(:get_and_update, data, index, next) when is_list(data) do
-    get_and_update_at(data, index, next, [])
+  defp at(:get_and_update, data, index, default, next) when is_list(data) do
+    get_and_update_at(data, index, default, next, [])
   end
 
-  defp at(_op, data, _index, _next) do
-    raise "Access.at/1 expected a list, got: #{inspect(data)}"
+  defp at(_op, data, _index, _default, _next) do
+    raise "Access.at/2 expected a list, got: #{inspect(data)}"
   end
 
-  defp get_and_update_at([head | rest], 0, next, updates) do
+  defp get_and_update_at([head | rest], 0, _default, next, updates) do
     case next.(head) do
       {get, update} -> {get, :lists.reverse([update | updates], rest)}
       :pop -> {head, :lists.reverse(updates, rest)}
     end
   end
 
-  defp get_and_update_at([_ | _] = list, index, next, updates) when index < 0 do
+  defp get_and_update_at([_ | _] = list, index, default, next, updates) when index < 0 do
     list_length = length(list)
 
     if list_length + index >= 0 do
-      get_and_update_at(list, list_length + index, next, updates)
+      get_and_update_at(list, list_length + index, default, next, updates)
     else
-      {nil, list}
+      {default, list}
     end
   end
 
-  defp get_and_update_at([head | rest], index, next, updates) when index > 0 do
-    get_and_update_at(rest, index - 1, next, [head | updates])
+  defp get_and_update_at([head | rest], index, default, next, updates) when index > 0 do
+    get_and_update_at(rest, index - 1, default, next, [head | updates])
   end
 
-  defp get_and_update_at([], _index, _next, updates) do
-    {nil, :lists.reverse(updates)}
+  defp get_and_update_at([], _index, default, _next, updates) do
+    {default, :lists.reverse(updates)}
```

Comment:
Hrm. In key/2, we will call “next” with the provided default. However here it doesn’t make sense, because we would need to fill in the list until the index. But this means the behavior is inconsistent within functions and within at/2 itself. Get is called with the default, get_and_update isn’t. Not sure how to proceed. Maybe we shouldn’t add default here after all? 🤔

Reply:
Honestly I got a bit lost in all the steps of `get_and_update_at`, so I made a stab at it and figured that the tests or the PR would catch anything I missed. I'll try to digest this after a break, though I suspect I might not have much to add. Happy to skip this if you think it's best :)

Reply:
The patch was perfect, the issue is rather a conceptual one. Unfortunately I don't think there is an easy answer, sorry for leading you in the wrong direction. We should still be fine for `at!` tho!